### PR TITLE
Modals support

### PR DESCRIPTION
**Pull request by jano42**
The aim of this pull request is to make blockly 100% integrable in modals (like with Bootstrap (http://getbootstrap.com/)

What changes ?
 * blockly can be used **in modals**
 * toolbox and workspace are placed in a master container (simple div) to allow **better toolbox positioning** (#141)
 * toolbox is placed next to injection container
 * toolbox of now placed correctly (fix bugs when the page scroll)
 * a click on the svg area don't make the container to grow up (simple css fix)
 * dropdown and tooltip now have a correct z-index (allow to be used in bootstrap modal)
 * toolbox RTL is working

Fix bugs
* #141 
* #181

### CODE REVIEW

**Comment by jano42** [comment]
@carlosperate : I've created the pr from scratch. It is simpler and easy to understand (I expect)

**Comment by carlosperate** [comment]
Could you provide a simple modal example to be able to test it?

**Comment by jano42** [comment]
This is a simple playground adaptation, to show a bootstrap3 modal dialog

[playground-bootstrap.zip](https://github.com/google/blockly/files/104861/playground-bootstrap.zip)
*(just place the contained html file, next to playground.html)*

**Comment by jano42** [comment]
@carlosperate is it good for you ? Have you some remarks or observe any problems ?

**Comment by carlosperate** [comment]
Sorry, I've been swamped with work and other projects, haven't had a chance to really look at it.

**Comment by jano42** [comment]
@NeilFraser : have a chance to merge ?

**Comment by jano42** [comment]
I signed it!

2016-04-08 15:15 GMT+02:00 googlebot <notifications@github.com>:

**Comment by ctrowat** [comment]
Any chance of this getting merged in soon?

**Comment by Immortalin** [comment]
@NeilFraser?

**Comment by jano42** [comment]
@NeilFraser?

**Comment by carlosperate** [comment]
@jano42 Out of curisity, did you build blockly using windows? Have you encountered issues where the `qqq.json` and `en.json` come up empty after a build?

**Comment by jano42** [comment]
Yes exactly ! I got  this issue. To solve it, I revert the  "en.json" (from master banch) and I build it a new time. Don't know exactly what happens.

**Comment by brunodunbar** [comment]
Is there any estimate when this will be integrated?

File: core/css.js [hunk 1]
```diff
@@ -135,12 +135,23 @@ Blockly.Css.CONTENT = [
     'background-color: #fff;',
     'outline: none;',
     'overflow: hidden;',  /* IE overflows by default. */
+	'display: block;',
```
Comment (NeilFraser) [inline core/css.js]: Spaces not tab.

File: core/css.js [hunk 2]
```diff
@@ -135,12 +135,23 @@ Blockly.Css.CONTENT = [
...
     'overflow: hidden;',  /* IE overflows by default. */
+	'display: block;', 
   '}',
 
   '.blocklyWidgetDiv {',
     'display: none;',
     'position: absolute;',
-    'z-index: 999;',
+    'z-index: 99999;', /* big value for bootstrap3 compatibility */
+  '}',
```
Comment (NeilFraser) [inline core/css.js]: Blank line between CSS blocks (here and 4 lines down).

File: core/inject.js [hunk 3]
```diff
@@ -50,8 +50,9 @@ Blockly.inject = function(container, opt_options) {
   if (!goog.dom.contains(document, container)) {
     throw 'Error: container is not in current document.';
   }
-  var options = new Blockly.Options(opt_options || {});
-  var svg = Blockly.createDom_(container, options);
+  var options = new Blockly.Options(opt_options || {});
```
Comment (NeilFraser) [inline core/inject.js]: Trailing whitespace on this line messes up the diff.

File: core/inject.js [hunk 4]
```diff
@@ -50,8 +50,9 @@ Blockly.inject = function(container, opt_options) {
   if (!goog.dom.contains(document, container)) {
     throw 'Error: container is not in current document.';
   }
-  var options = new Blockly.Options(opt_options || {});
-  var svg = Blockly.createDom_(container, options);
+  var options = new Blockly.Options(opt_options || {});  
+  var subContainer = goog.dom.createDom('div', 'injectionDiv');
+  container.appendChild(subContainer);   var svg = Blockly.createDom_(subContainer, options);
```
Comment (NeilFraser) [inline core/inject.js]: Missing line break here.

File: core/toolbox.js [hunk 5]
```diff
@@ -231,30 +232,28 @@ Blockly.Toolbox.prototype.position = function() {
   var svgPosition = goog.style.getPageOffset(svg);
   var svgSize = Blockly.svgSize(svg);
   if (this.horizontalLayout_) {
-    treeDiv.style.left = svgPosition.x + 'px';
+    treeDiv.style.left = '0px';
```
Comment (NeilFraser) [inline core/toolbox.js]: Just 0, no need for '0px'.  Here and below 4 more times.

File: core/toolbox.js [hunk 6]
```diff
@@ -231,30 +232,28 @@ Blockly.Toolbox.prototype.position = function() {
...
     treeDiv.style.height = 'auto';
     treeDiv.style.width = svgSize.width + 'px';
     this.height = treeDiv.offsetHeight;
     if (this.toolboxPosition == Blockly.TOOLBOX_AT_TOP) {  // Top
-      treeDiv.style.top = svgPosition.y + 'px';
+      treeDiv.style.top ='0px';
```
Comment (NeilFraser) [inline core/toolbox.js]: Missing space after = sign.

File: core/toolbox.js [hunk 7]
```diff
@@ -231,30 +232,28 @@ Blockly.Toolbox.prototype.position = function() {
...
     } else {  // Left
-      treeDiv.style.left = svgPosition.x + 'px';
+      treeDiv.style.left = '0px';
     }
     treeDiv.style.height = svgSize.height + 'px';
-    treeDiv.style.top = svgPosition.y + 'px';
+    //treeDiv.style.top = svgPosition.y + 'px';
     this.width = treeDiv.offsetWidth;
-    if (this.toolboxPosition == Blockly.TOOLBOX_AT_LEFT) {
+    /*if (this.toolboxPosition == Blockly.TOOLBOX_AT_LEFT) {
```
Comment (NeilFraser) [inline core/toolbox.js]: Just delete unused code rather than commenting it out.  Here and two lines up.

**Comment by NeilFraser** [comment]
I've tested this PR to death and I can find no bugs.  Impressive (and totally unexpected).
There are a few style comments.  Other than than, can you redirect this PR to the 'develop' branch (instead of 'master'), and revert the language and compiled files.  LGTM.
In my opinion this wholly supplants #492.

**Comment by jano42** [comment]
@NeilFraser : I've merge 'develop' branch into my PR. I don't know what you meen by "redirect this PR to the 'develop' branch". (I'm relatively new with git, and I don't know if I can do something else than merging).

**Comment by carlosperate** [comment]
When you create a PR you have select which branch you are targeting, in this case I believe you would have to create a new PR to target the develop branch instead of master.
At this point I'd recommend (in your local git repository) to create a new branch from develop, merge squash your changes there, and submit that as a new PR.

**Comment by jano42** [comment]
@carlosperate : Thanks for tips
@NeilFraser : OK a new PR #512 is ready to merge into develop branch

:+1:

**Comment by NeilFraser** [comment]
#512 has been merged.  Thanks!  Closing this PR.
👍

**Pull request closed without merging**